File: backend/app/services/goal_service.py
from datetime import datetime
from sqlalchemy.orm import Session

from app.models.goal import Goal
from app.schemas.goal import GoalCreate


def create_goal(
    db: Session,
    goal: GoalCreate,
    user_id: int
):
    new_goal = Goal(
        title=goal.title,
        description=goal.description,
        category=goal.category,
        target_date=goal.target_date,
        user_id=user_id
    )

    db.add(new_goal)
    db.commit()
    db.refresh(new_goal)

    return new_goal


def get_all_goals(
    db: Session,
    user_id: int
):
    return (
        db.query(Goal)
        .filter(
            Goal.user_id == user_id,
            Goal.is_archived == False
        )
        .all()
    )


def get_goal(
    db: Session,
    goal_id: int,
    user_id: int
):
    return (
        db.query(Goal)
        .filter(
            Goal.id == goal_id,
            Goal.user_id == user_id
        )
        .first()
    )


def update_goal(
    db: Session,
    goal_id: int,
    goal: GoalCreate,
    user_id: int
):
    existing_goal = get_goal(
        db,
        goal_id,
        user_id
    )

    if not existing_goal:
        return None

    existing_goal.title = goal.title
    existing_goal.description = goal.description
    existing_goal.category = goal.category
    existing_goal.target_date = goal.target_date

    db.commit()
    db.refresh(existing_goal)

    return existing_goal


def delete_goal(
    db: Session,
    goal_id: int,
    user_id: int
):
    goal = get_goal(
        db,
        goal_id,
        user_id
    )

    if not goal:
        return None

    db.delete(goal)
    db.commit()

    return goal


def archive_goal(
    db: Session,
    goal_id: int,
    user_id: int
):
    goal = get_goal(
        db,
        goal_id,
        user_id
    )

    if not goal:
        return None

    goal.is_archived = True

    db.commit()
    db.refresh(goal)

    return goal


def restore_goal(
    db: Session,
    goal_id: int,
    user_id: int
):
    goal = get_goal(
        db,
        goal_id,
        user_id
    )

    if not goal:
        return None

    goal.is_archived = False

    db.commit()
    db.refresh(goal)

    return goal


def update_progress(
    db: Session,
    goal_id: int,
    progress: int,
    user_id: int
):
    goal = get_goal(
        db,
        goal_id,
        user_id
    )

    if not goal:
        return None

    goal.progress = max(0, min(progress, 100))

    if goal.progress == 100:
        goal.status = "Completed"
    else:
        goal.status = "Active"

    goal.updated_at = datetime.utcnow()

    db.commit()
    db.refresh(goal)

    return goal

File: frontend/services/goal.service.ts
import { api } from "@/lib/api";
import type { GoalResponse, CreateGoalPayload, UpdateGoalPayload, GoalStatus } from "@/types/goal";

// Wire shape from app/schemas/goal.py -- snake_case target_date/is_archived
// instead of camelCase targetDate/archived, and no single generic
// "archived" flag to PATCH (there are dedicated /archive and /restore
// routes instead).
interface BackendGoal {
  id: number;
  title: string;
  description: string;
  category: string;
  target_date: string | null;
  progress: number;
  status: string;
  is_archived: boolean;
}

function toGoalResponse(g: BackendGoal): GoalResponse {
  const status: GoalStatus = g.is_archived
    ? "Archived"
    : g.status === "Completed"
      ? "Completed"
      : "Active";
  return {
    id: g.id,
    title: g.title,
    description: g.description,
    category: g.category,
    targetDate: g.target_date ?? "",
    progress: g.progress,
    status,
    archived: g.is_archived,
  };
}

export async function getGoals(): Promise<GoalResponse[]> {
  const res = await api.get<BackendGoal[]>("/goals");
  return res.map(toGoalResponse);
}

export async function createGoal(payload: CreateGoalPayload): Promise<GoalResponse> {
  const res = await api.post<BackendGoal>("/goals", {
    title: payload.title,
    description: payload.description,
    category: payload.category,
    target_date: payload.targetDate || undefined,
  });
  return toGoalResponse(res);
}

export async function updateGoal(id: number, payload: UpdateGoalPayload): Promise<GoalResponse> {
  const body: Record<string, unknown> = {};
  if (payload.title !== undefined) body.title = payload.title;
  if (payload.description !== undefined) body.description = payload.description;
  if (payload.category !== undefined) body.category = payload.category;
  if (payload.targetDate !== undefined) body.target_date = payload.targetDate || undefined;

  let res = await api.patch<BackendGoal>(`/goals/${id}`, body);

  // `status`/`archived` aren't part of the generic PATCH -- they have
  // dedicated routes.
  if (payload.progress !== undefined) {
    res = await api.put<BackendGoal>(`/goals/${id}/progress/${payload.progress}`);
  }
  if (payload.archived === true) {
    await api.put<void>(`/goals/${id}/archive`);
    res = { ...res, is_archived: true };
  } else if (payload.archived === false) {
    await api.put<void>(`/goals/${id}/restore`);
    res = { ...res, is_archived: false };
  }

  return toGoalResponse(res);
}

export async function updateGoalProgress(id: number, progress: number): Promise<GoalResponse> {
  const res = await api.put<BackendGoal>(`/goals/${id}/progress/${progress}`);
  return toGoalResponse(res);
}

export async function archiveGoal(id: number): Promise<GoalResponse> {
  await api.put<void>(`/goals/${id}/archive`);
  const res = await api.get<BackendGoal>(`/goals/${id}`);
  return toGoalResponse(res);
}

export function deleteGoal(id: number): Promise<void> {
  return api.delete<void>(`/goals/${id}`);
}
